[DO NOT MERGE] Entity Prototype for SDK Specification

jsuereth opened the pull request:
Replaces #6855
 
 Implementation of [OTEP#264](https://github.com/open-telemetry/opentelemetry-specification/blob/main/oteps/entities/0264-resource-and-entities.md)
 
 - Updates OTLP Resource Marshaler code for OTLP 1.7 (which includes entity ref).
 - Puts new entity code in `io.opentelemetry.sdk.resources.internal`
 - `Resource` and `ResourceBuilder` do NOT follow other data class conventions in the SDK (of being a pure interface), so we are
    unable to have this code anywhere but inside the SDK to make everything work.
 -  Adds entity merge algorithm, partitioned into three key components.
 - Updates `toString` tests to be more flexible
 - Creates an `entity` incubator package
   - Adds bare minimum `EntityProvider`
   - Adds minimal `EntityDetector`
   - Adds two implementation of `EntityDetector` for `service` and `telemetry.sdk`

## Context

- [Entity Data Model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/entities/data-model.md)
-  *updated* [Resource Data Model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/data-model.md)
- [Design](https://github.com/open-telemetry/opentelemetry-specification/blob/main/oteps/entities/0264-resource-and-entities.md)
 
##  TODOs
 
 - [x] Figure out where to put `Entity` classes
 - [x] Figure out what to do about `Resource` in the SDK.  Bincompat prevents many good design options.
 - [x] Update `telemetry.sdk` and `service` resource attributes to be entities *and* include a `SchemaURL`.
 - [x] `EntityDetector` mechanism and wiring
 - [x] `EntityProvider` or `ResourceProvider` in the SDK, similar to `TraceProvider`, `MeterProvider`, etc.
 
The current design represents a "middle ground" for binary compatibility given the state of `Resource` today in the SDK.

jsuereth committed:
Reboot the entity prototype

- Create Entity/EntityBuilder class in internal package
- Update serialization code

Need to discuss how to work with Resource going forward.

jsuereth committed:
Add base entity class, merge logic and test

jsuereth committed:
Move Entity and EntityBuilder to be pure interfaces

jsuereth committed:
Add more merge helper methods and tests

jsuereth committed:
Added test for merging raw attributes

jsuereth committed:
Cleanups and reverting silly changes

jsuereth committed:
fix minor style issue

jsuereth committed:
Actually wire entities into resource

- Remove ResourceWithEntity, no way to keep bincompat that way
- Update toString tests to be a lot more forgiving, but preserve intent

jsuereth committed:
Fix typo

jsuereth committed:
Remove all publicly accessible references to Entity from stable packages

jsuereth committed:
Add incubating EntityProvider API for specification work

jsuereth committed:
Fix refactoring that broke reflection.

- fix reflective method lookup
- also add unit test in the right project.

jsuereth committed:
Fix schema url and test, now that we proved it works

jsuereth committed:
Initial cut at moving Entity SDK to have correpsonding API.

- Creates an API matching EntityProvider OTEP
- Updates EntityDetector to be ResourceDetector
- Creates synchronous "update resource with this entity" method.
- Adds wiring for incubating API + incubating SDK

Note: Still have some thoughts and ideas about a formal API, this may still change.

jsuereth committed:
Further simplify the API and SDK for entities

- Remove `Entity` and only expose `EntityBuilder` similar to how we do `LogRecord`s.
- Have methods on `Resource` which determine if you're attaching the entity, simplify "remove" or just ignore for now.

jsuereth committed:
Wire ResourceProvider in end-to-end test with new incubating ExtendedOpenTelemetry API.

- Update shared state in SDK to use Supplier<Resource> instead of Resource
- Add helper utils to expose private methods to supply the supplier
- Create new ExtendedOpenTelemetry* API/SDK and end-to-end test.

jsuereth committed:
Fix javadoc issues

jack-berg commented on sdk/common/src/main/java/io/opentelemetry/sdk/resources/internal/EntityUtil.java:
```diff
@@ -0,0 +1,295 @@
+      Collection<Entity> entities, AttributeKey<T> key) {
+    return entities.stream()
+        .anyMatch(
+            e -> e.getId().asMap().containsKey(key) || e.getDescription().asMap().containsKey(key));
```
There's a couple of places where we the code calls `Attributes.asMap().containsKey(key)`. Its much more efficient to call `Attributes.get(key) != null`.

jack-berg commented on sdk/common/src/main/java/io/opentelemetry/sdk/resources/internal/SdkEntity.java:
```diff
@@ -0,0 +1,54 @@
+   *
+   * @param entityType the entity type string of this entity.
+   */
+  public static final EntityBuilder builder(String entityType) {
```
No need to declare a static method final.

jack-berg commented on sdk/common/src/main/java/io/opentelemetry/sdk/resources/internal/SdkEntity.java:
```diff
@@ -0,0 +1,54 @@
+      String entityType, Attributes id, Attributes description, @Nullable String schemaUrl) {
+    AttributeCheckUtil.isValid(entityType);
+    AttributeCheckUtil.checkAttributes(id);
+    AttributeCheckUtil.checkAttributes(description);
```
Perform these checks earlier. `entityType` can be validated in `SdkEntity.builder(String entityType)`, and `id` / `description` can be validated in the setters in `EntityBuilder`.

jsuereth replied:
Done.

jack-berg commented on sdk/logs/src/main/java/io/opentelemetry/sdk/logs/LoggerSharedState.java:
```diff
@@ -18,28 +18,29 @@
 final class LoggerSharedState {
   private final Object lock = new Object();
-  private final Resource resource;
+  private final Supplier<Resource> resourceSupplier;
```
👍

jack-berg commented on sdk/logs/src/main/java/io/opentelemetry/sdk/logs/LoggerSharedState.java:
```diff
@@ -18,28 +18,29 @@
 
-  Resource getResource() {
-    return resource;
+  // This is used in a test, and must be public for it.
```
This is incorrect. All the tests calling this are in the same package so package private will suffice.

jsuereth replied:
Ah, yes I think I removed this.

jsuereth replied:
Resolved

jack-berg commented on sdk/logs/src/main/java/io/opentelemetry/sdk/logs/SdkLoggerProviderBuilder.java:
```diff
@@ -69,6 +71,23 @@ public SdkLoggerProviderBuilder addResource(Resource resource) {
+   * SdkLoggerProviderUtil#setResourceSupplier(SdkLoggerProviderBuilder, Supplier)}.
+   *
+   * @param supplier The supplier of {@link Resource}.
+   * @since 1.X.0
```
Remove - i'll go back and add appropriate `@since` tags on all new methods as part of the release process. 
```suggestion
```

jack-berg commented on sdk/logs/src/main/java/io/opentelemetry/sdk/logs/SdkLoggerProviderBuilder.java:
```diff
@@ -69,6 +71,23 @@ public SdkLoggerProviderBuilder addResource(Resource resource) {
+   * @param supplier The supplier of {@link Resource}.
+   * @since 1.X.0
+   */
+  SdkLoggerProviderBuilder setResourceSupplier(Supplier<Resource> supplier) {
```
I don't see these new `Sdk{Signal}ProviderBuilder#setResourceSupplier` methods being exercised. How do these fit into the puzzle?

jack-berg reviewed (commented):
Left a variety of comments, but the only material ones are about naming.

jsuereth commented on api/incubator/src/main/java/io/opentelemetry/api/incubator/entities/EntityBuilder.java:
```diff
@@ -0,0 +1,55 @@
+   * @param f A {@link Consumer} which builds the descriptive attributes.
+   * @return this
+   */
+  EntityBuilder withDescription(Consumer<AttributesBuilder> f);
```
That's a great suggestion!

For reference: I'm not married to the specifics of the builders in this API, this PR is mostly focused on "shape" and "behavior".

jsuereth replied:
Updated

jsuereth commented on api/incubator/src/main/java/io/opentelemetry/api/incubator/entities/EntityBuilder.java:
```diff
@@ -0,0 +1,55 @@
+  EntityBuilder withId(Consumer<AttributesBuilder> f);
+
+  /** Emits the current entity. */
+  void emit();
```
This is anticipation of having an EntityBuilder that just emits Entity without attaching to active resource.

I debated two things:

- Having `EntityBuilder` have two methods:  `emit` and `attach` where this PR would only have `attach`.
- Having the `Resource` method you call to get the builder be named `attach` or `emit`.

I was planning to talk about this awkwardness in the next Entities SIG, but agree with only "attach" then this should be "attach".

jsuereth commented on api/incubator/src/main/java/io/opentelemetry/api/incubator/entities/ExtendedOpenTelemetry.java:
```diff
@@ -0,0 +1,21 @@
+  }
+
+  /** Returns the {@link Resource} that telemetry from this {@link OpenTelemetry} uses. */
+  default Resource getResource() {
```
Sounds good, will remove

jsuereth commented on api/incubator/src/main/java/io/opentelemetry/api/incubator/entities/Resource.java:
```diff
@@ -0,0 +1,25 @@
+package io.opentelemetry.api.incubator.entities;
+
+/** The active resource for which Telemetry is being generated. */
+public interface Resource {
```
Agreed, and this is the another thing we want to discuss in the Entities SIG.  I'm still working on a new set of names for this here.

It's possible we move to something like:

```
interface EntityProvider {
  EntityBuilder attachToCurrentResource(String entityType);
  EntityBuilder report(String entityType);
}
```

With the caveat - I haven't come up with a set of names I'm happy with.

jsuereth commented on docs/apidiffs/current_vs_latest/opentelemetry-sdk-common.txt:
```diff
@@ -1,2 +1,5 @@
\ No newline at end of file
+***  MODIFIED CLASS: PUBLIC ABSTRACT io.opentelemetry.sdk.resources.Resource  (not serializable)
+	===  CLASS FILE FORMAT VERSION: 52.0 <- 52.0
+	+++  NEW METHOD: PUBLIC(+) STATIC(+) io.opentelemetry.sdk.resources.Resource create(io.opentelemetry.api.common.Attributes, java.lang.String, java.util.Collection<io.opentelemetry.sdk.resources.internal.Entity>)
```
AH, yes, I need to wrap this in reflective access. Good catch!

jsuereth committed:
Update exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/otlp/EntityRefMarshaler.java

Co-authored-by: jack-berg <34418638+jack-berg@users.noreply.github.com>

jsuereth committed:
Update sdk/logs/src/main/java/io/opentelemetry/sdk/logs/SdkLoggerProviderBuilder.java

Co-authored-by: jack-berg <34418638+jack-berg@users.noreply.github.com>

jsuereth committed:
Fixes from review.

- Hide more methods (no incubating methods in public APIs that are not internal)
- Update builder API to use attributes directly
- Use static instances in Noop API

jsuereth committed:
Remove methods we don't need yet.

jsuereth committed:
Remove redundant visibility.

jsuereth commented on api/incubator/src/main/java/io/opentelemetry/api/incubator/entities/NoopResource.java:
```diff
@@ -0,0 +1,19 @@
+
+  @Override
+  public EntityBuilder attachEntity(String entityType) {
+    return new NoopEntityBuilder();
```
Fixed

jsuereth committed:
More cleanups from review.

jsuereth committed:
Simplify prototype of SdkResource.

jsuereth committed:
Fix typo.

jsuereth commented on sdk/common/src/main/java/io/opentelemetry/sdk/resources/internal/EntityUtil.java:
```diff
@@ -0,0 +1,295 @@
+    Collection<Entity> entities = EntityUtil.mergeEntities(getEntities(base), getEntities(next));
+    RawAttributeMergeResult attributeResult =
+        EntityUtil.mergeRawAttributes(getRawAttributes(base), getRawAttributes(next), entities);
+    // Remove entiites that are conflicting with raw attributes, and therefore in an unknown state.
```
should be fixed

jsuereth committed:
More cleanups from review.

jsuereth commented on sdk/common/src/main/java/io/opentelemetry/sdk/resources/internal/EntityUtil.java:
```diff
@@ -0,0 +1,295 @@
+   * @param additional Additional entities to merge with base set.
+   * @return A new set of entities with no duplicate types.
+   */
+  public static final Collection<Entity> mergeEntities(
```
Fixed

jsuereth committed:
Add more tests for ResourceProvider.

jsuereth committed:
Fix stylecheck.

jsuereth committed:
More checkstyle fixes.

jsuereth committed:
Rename prototype to more closely match OTEP 4316.

jsuereth committed:
Finish remove implementation and listeners.

jsuereth committed:
Remove assertjs helpers, update javadoc.

jsuereth committed:
Fix entity javadoc.

jsuereth commented on sdk/common/src/main/java/io/opentelemetry/sdk/resources/internal/Entity.java:
```diff
@@ -0,0 +1,74 @@
+  /**
+   * Returns a map of attributes that identify the entity.
+   *
+   * @return a map of attributes.
```
Fixed

jsuereth committed:
Fix tests.

jsuereth commented on sdk/common/src/main/java/io/opentelemetry/sdk/resources/internal/EntityUtil.java:
```diff
@@ -0,0 +1,295 @@
+        method.setAccessible(true);
+        method.invoke(rb, e);
+      }
+    } catch (NoSuchMethodException nme) {
```
Fixed

jsuereth commented on exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/otlp/EntityRefMarshaler.java:
```diff
@@ -0,0 +1,83 @@
+    return new EntityRefMarshaler(
+        schemaUrlUtf8,
+        e.getType().getBytes(StandardCharsets.UTF_8),
+        e.getId().asMap().keySet().stream()
```
Also, if it wasn't apparent for now - This marshaller is only used ONCE per resource instance.  So unless resource is being updated frequently (which we do not expect), then this cost is not paid often, if more than once.

jsuereth committed:
Tweak incubator SDK to support asynchronous configuratio and use EntityListener pattern from OTEP.

jsuereth commented:
Closing this in favor of a new prototype
